### ISSUE TITLE
Superresolution GAN: replace waifu2x

### ISSUE THREAD

**Issue by gwern** [tensorfork/tensorfork#25]
Superresolution is a key part of our workflows: many images, particularly if cropped down to a face or figure, are just not big enough to train natively at a resolution like 512px or 1024px. Padding them with black bars doesn't help, and regular upscaling algorithms do a poor job and create artifacts which damage all users.

Currently, we use `waifu2x` which is specialized for anime/illustrations. However, waifu2x is showing its age (original release ~2014!): it's implemented in *Torch*, which is bound to an old CUDA/CuDNN library and it will break soon. The maintainer has no active plans to reimplement it in Tensorflow/PyTorch. Further, waifu2x is quite slow, requiring many seconds to upscale larger images. Finally, waifu2x has not delivered SOTA results for half a decade, and is long obsolete. (If you go beyond 2x upscaling, the artifacts become especially obvious.) The reason we don't use another superresolution like ESRGAN is that they are always trained exclusively on photographic materials, which probably means they will deliver worse results than waifu2x on our datasets.

This presents an opportunity to train & release a large powerful SOTA anime/illustration-oriented superresolution NN. It'll improve our upscales, if good enough may let us relax size requirements and allow >2x upscales.

What superresolution? Unclear. I don't know offhand, and [Papers With Code](https://paperswithcode.com/task/super-resolution/latest) only has a leaderboard of 2 (FALSR-A & RCAN). Its [specialized leaderboard](https://paperswithcode.com/task/image-super-resolution) suggests [PFF](https://github.com/aimerykong/predictive-filter-flow/tree/master/PFF_image) (2018) or [SAN](https://github.com/daitao/SAN) (2019). Both appear to use PyTorch. A third competitor is the more well known [ESRGAN](https://paperswithcode.com/paper/esrgan-enhanced-super-resolution-generative#code) which does have TensorFlow implementations & [a Keras implementation](https://github.com/idealo/image-super-resolution) (unclear how mature any of them are).

On the positive side, it appears that since superresolution is such a strongly-supervised task, we probably don't even need the TPU pods and can just train locally on GPU, so PyTorch is totally an option.

**Comment by Skylion007** [comment]
ESRGAN would be what I'd recommend on using due to its simplicity and plethora of implementations.

**Comment by gwern** [comment]
[SAN_pytorch](https://github.com/Yukariin/SAN_pytorch) may be an option if its [runtime performance](https://github.com/Yukariin/SAN_pytorch/issues/3) improves.

**Comment by victorca25** [comment]
> The reason we don't use another superresolution like ESRGAN is that they are always trained exclusively on photographic materials, which probably means they will deliver worse results than waifu2x on our datasets.

That is not correct. In the Game Upscale community we have people that specifically trained models focused on manga/anime. Look here: https://upscale.wiki/wiki/Model_Database

Some changes to the original ESRGAN training strategy were made in my fork, but it's the same architecture for all those models.

**Comment by gwern** [comment]
I've tried out [JoeyBallantine's ESRGAN fork](https://github.com/JoeyBallentine/ESRGAN) which has many anime-specific models (specifically, `4x_NMKD-YandereNeoXL_200k.pth`). Seems pretty good, and not too hard to get running.

**Comment by victorca25** [comment]
Just to clarify, the ESRGAN repos are only for inference of the models. There's also the Cupscale app that makes using the models in the database even easier, specially if you don't want to set up a python environment manually.

In order to train the models, like the ones in the database, use: https://github.com/victorca25/BasicSR. There are multiple different networks that can be trained, with different options for each case, like real images or anime/manga.

In the community there are also reduced sizes of the model that produce good results.
